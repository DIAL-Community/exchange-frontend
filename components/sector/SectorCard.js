import { useCallback } from 'react'
import { useIntl } from 'react-intl'
import Link from 'next/link'
import parse from 'html-react-parser'
import { FaXmark } from 'react-icons/fa6'
import { DisplayType } from '../utils/constants'
import { isValidFn } from '../utils/utilities'

const SectorCard = ({ displayType, index, sector, dismissHandler }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const displayLargeCard = () =>
    <div className={`px-4 py-6 rounded-lg min-h-[8rem] ${index % 2 === 0 && 'bg-dial-violet'}`}>
      <div className='flex flex-col lg:flex-row gap-x-6 gap-y-3'>
        <div className='flex flex-col gap-y-3 max-w-3xl lg:w-10/12'>
          <div className='text-lg font-semibold text-dial-plum'>
            {sector.name}
          </div>
          <div className='line-clamp-4 text-dial-stratos'>
            {sector?.sectorDescription && parse(sector?.sectorDescription.description)}
          </div>
          <div className='flex gap-x-2 text-dial-stratos'>
            <div className='text-sm'>
              {format('ui.dataset.header')} ({sector.datasets?.length ?? 0})
            </div>
            <div className='border-r border-dial-slate-400' />
            <div className='text-sm'>
              {format('ui.organization.header')} ({sector.organizations?.length ?? 0})
            </div>
            <div className='border-r border-dial-slate-400' />
            <div className='text-sm'>
              {format('ui.product.header')} ({sector.products?.length ?? 0})
            </div>
            <div className='border-r border-dial-slate-400' />
            <div className='text-sm'>
              {format('ui.project.header')} ({sector.projects?.length ?? 0})
            </div>
          </div>
        </div>
      </div>
    </div>

  const displaySmallCard = () =>
    <div className='rounded-lg bg-gradient-to-r from-sector-bg-light to-sector-bg h-16'>
      <div className='flex flex-row gap-x-3 px-6 h-full'>
        <div className='text-sm font-semibold text-dial-plum my-auto'>
          {sector.name}
        </div>
      </div>
    </div>

  return (
    <div className='relative'>
      <Link href={`/sectors/${sector.slug}`}>
        {displayType === DisplayType.LARGE_CARD && displayLargeCard()}
        {displayType === DisplayType.SMALL_CARD && displaySmallCard()}
      </Link>
      { isValidFn(dismissHandler) &&
        <button type='button' className='absolute top-2 right-2'>
          <FaXmark size='1rem' className='text-dial-plum' onClick={dismissHandler} />
        </button>
      }
    </div>
  )
}

export default SectorCard
