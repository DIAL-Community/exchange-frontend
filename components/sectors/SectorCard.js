import Link from 'next/link'
import { useIntl } from 'react-intl'

const SectorCard = ({ sector, listType }) => {
  const { formatMessage } = useIntl()
  const format = (id) => formatMessage({ id })
  return (
    <>
      {listType === 'list'
        ? (
          <div className='border-3 border-transparent hover:border-dial-yellow text-button-gray hover:text-dial-yellow cursor-pointer'>
            <div className='border border-dial-gray hover:border-transparent shadow-sm hover:shadow-lg'>
              <div className='flex justify-between my-4 px-4'>
                <div className='inline-block card-title truncate card-link-text text-button-gray'>
                  {sector.name}
                </div>
              </div>
            </div>
          </div>
          )
        : (
          <div>Card View</div>
          )
        }
    </>
  )
}

export default SectorCard
