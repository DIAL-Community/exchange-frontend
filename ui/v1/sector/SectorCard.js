import { useCallback } from 'react'
import { useIntl } from 'react-intl'
import Link from 'next/link'
import parse from 'html-react-parser'
import { IoClose } from 'react-icons/io5'
import { DisplayType, REBRAND_BASE_PATH } from '../utils/constants'

const SectorCard = ({ displayType, index, sector, dismissCardHandler }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const displayLargeCard = () =>
    <div className={`px-4 py-6 rounded-lg min-h-[13.5rem] ${index % 2 === 0 && 'bg-dial-violet'}`}>
      <div className='flex flex-col lg:flex-row gap-x-6 gap-y-3'>
        <div className='flex flex-col gap-y-3 max-w-3xl lg:w-10/12'>
          <div className='text-lg font-semibold text-dial-plum'>
            {sector.name}
          </div>
          <div className='line-clamp-4 text-dial-stratos'>
            {sector?.sectorDescription && parse(sector?.sectorDescription.description)}
          </div>
          <div className='flex gap-x-2 text-dial-stratos'>
            <div className='text-sm'>
              {format('ui.sector.header')} ({sector.sectors?.length ?? 0})
            </div>
            <div className='border border-r text-dial-stratos-300' />
            <div className='text-sm'>
              {format('ui.country.header')} ({sector.countries?.length ?? 0})
            </div>
            <div className='border border-r text-dial-stratos-300' />
            <div className='text-sm'>
              {format('ui.project.header')} ({sector.projects?.length ?? 0})
            </div>
          </div>
        </div>
      </div>
    </div>

  const displaySmallCard = () =>
    <div className='rounded-lg bg-gradient-to-r from-sector-bg-light to-sector-bg h-16'>
      <div className='flex flex-row gap-x-3 px-6 h-full'>
        <div className='text-sm font-semibold text-dial-plum my-auto'>
          {sector.name}
        </div>
      </div>
    </div>

  return (
    <div className='relative'>
      <Link href={`${REBRAND_BASE_PATH}/sectors/${sector.slug}`}>
        {displayType === DisplayType.LARGE_CARD && displayLargeCard()}
        {displayType === DisplayType.SMALL_CARD && displaySmallCard()}
      </Link>
      {dismissCardHandler && {}.toString.call(dismissCardHandler) === '[object Function]' &&
        <button className='absolute p-2 top-0 right-0 text-dial-sapphire'>
          <IoClose size='1rem' onClick={dismissCardHandler} />
        </button>
      }
    </div>
  )
}

export default SectorCard
