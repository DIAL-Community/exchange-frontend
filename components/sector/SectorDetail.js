import { useRef } from 'react'
import { useApolloClient, useQuery } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'
import Breadcrumb from '../shared/Breadcrumb'
import { handleLoadingQuery, handleMissingData, handleQueryError } from '../shared/GraphQueryHandler'
import { SECTOR_DETAIL_QUERY, SECTOR_POLICY_QUERY } from '../shared/query/sector'
import { fetchOperationPolicies } from '../utils/policy'
import SectorDetailLeft from './SectorDetailLeft'
import SectorDetailRight from './SectorDetailRight'

const SectorDetail = ({ slug }) => {
  const scrollRef = useRef(null)
  const client = useApolloClient()

  const policies = fetchOperationPolicies(
    client,
    SECTOR_POLICY_QUERY,
    ['editing', 'deleting']
  )

  const editingAllowed = policies['editing']
  const deletingAllowed = policies['deleting']

  const { loading, error, data } = useQuery(SECTOR_DETAIL_QUERY, {
    variables: { slug },
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  if (loading) {
    return handleLoadingQuery()
  } else if (error) {
    return handleQueryError(error)
  } else if (!data?.sector) {
    return handleMissingData()
  }

  const { sector } = data

  const slugNameMapping = (() => {
    const map = {}
    map[sector.slug] = sector.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping}/>
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3'>
          <SectorDetailLeft scrollRef={scrollRef} sector={sector} />
        </div>
        <div className='lg:basis-2/3'>
          <SectorDetailRight
            ref={scrollRef}
            sector={sector}
            editingAllowed={editingAllowed}
            deletingAllowed={deletingAllowed}
          />
        </div>
      </div>
    </div>
  )
}

export default SectorDetail
