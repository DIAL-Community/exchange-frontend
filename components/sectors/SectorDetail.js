import { useCallback } from 'react'
import { useIntl } from 'react-intl'
import { useQuery } from '@apollo/client'
import { useRouter } from 'next/router'
import Dialog from '../shared/Dialog'
import Checkbox from '../shared/Checkbox'
import { SECTOR_SEARCH_QUERY } from '../../queries/sector'
import { getLanguageOptions } from '../../lib/utilities'

const labelTextBlue = 'text-dial-blue'
const labelTextGray = 'text-button-gray'

const SectorDetail = ({ sector, isOpen, onClose }) => {
  const { locale } = useRouter()
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const { data } = useQuery(SECTOR_SEARCH_QUERY, {
    variables: { search: '', locale }
  })

  const parentSector = data?.sectors?.find(({ id }) => id === sector?.parentSectorId?.toString())
  const sectorLocale = getLanguageOptions(format).find(({ value } ) => value === sector?.locale)

  return (
    <Dialog
      closeButton
      isOpen={isOpen}
      onClose={onClose}
    >
      <div className='flex flex-col gap-3' data-testid='sector-detail'>
        <div className={`${labelTextBlue} text-2xl font-bold pb-4`}>
          <p data-testid='sector-name'>{`${format('sector.label')}: ${sector.name}`}</p>
        </div>
        <div className='flex flex-col gap-3'>
          {sector?.parentSectorId && (
            <div className='text-xl'>
              <p className={labelTextBlue}>{`${format('sector.parent-sector.label')}:`}</p>
              <p className={labelTextGray} data-testid='parent-sector'>{parentSector?.name}</p>
            </div>
          )}
          <div className='text-xl'>
            <p className={labelTextBlue}>{`${format('locale.label')}:`}</p>
            <p className={labelTextGray} data-testid='sector-locale'>{sectorLocale.label}</p>
          </div>
          <div className={'flex flex-row text-xl gap-x-2'}>
            <Checkbox value={sector.isDisplayable} disabled/>
            <p className={labelTextBlue}>{format('sector.is-displayable.label')}</p>
          </div>
        </div>
      </div>  
    </Dialog>
  )
}

export default SectorDetail
