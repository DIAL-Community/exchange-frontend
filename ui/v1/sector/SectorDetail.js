import { useRef } from 'react'
import { useQuery } from '@apollo/client'
import { SECTOR_DETAIL_QUERY } from '../shared/query/sector'
import Breadcrumb from '../shared/Breadcrumb'
import { Error, Loading, NotFound } from '../shared/FetchStatus'
import SectorDetailRight from './SectorDetailRight'
import SectorDetailLeft from './SectorDetailLeft'

const SectorDetail = ({ slug }) => {
  const scrollRef = useRef(null)
  const commentsSectionRef = useRef(null)

  const { loading, error, data } = useQuery(SECTOR_DETAIL_QUERY, {
    variables: { slug }
  })

  if (loading) {
    return <Loading />
  } else if (error) {
    return <Error />
  } else if (!data?.sector) {
    return <NotFound />
  }

  const { sector } = data

  const slugNameMapping = (() => {
    const map = {}
    map[sector.slug] = sector.name

    return map
  })()

  return (
    <div className='lg:px-8 xl:px-56 flex flex-col'>
      <div className='px-4 lg:px-6 py-4 bg-dial-violet text-dial-stratos ribbon-detail z-40'>
        <Breadcrumb slugNameMapping={slugNameMapping}/>
      </div>
      <div className='flex flex-col lg:flex-row gap-x-8'>
        <div className='lg:basis-1/3'>
          <SectorDetailLeft scrollRef={scrollRef} sector={sector} />
        </div>
        <div className='lg:basis-2/3'>
          <SectorDetailRight ref={scrollRef} commentsSectionRef={commentsSectionRef} sector={sector} />
        </div>
      </div>
    </div>
  )
}

export default SectorDetail
