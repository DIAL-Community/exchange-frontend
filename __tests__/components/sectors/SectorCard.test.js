import { mockRouterImplementation, mockSessionImplementation, render, waitForAllEffects } from '../../test-utils'
import CustomMockedProvider from '../../utils/CustomMockedProvider'
import SectorCard from '../../../components/sectors/SectorCard'
import { sectorWithParentSector } from './data/SectorDetail'

jest.mock('next/dist/client/router')
jest.mock('next-auth/client')

describe('Unit test for the SectorCard component', () => {
  const CARD_TEST_ID = 'sector-card'
  const EDIT_BUTTON_TEST_ID = 'edit-button'
  const DELETE_BUTTON_TEST_ID = 'delete-button'

  beforeAll(mockRouterImplementation)

  describe('Should match snapshot -', () => {
    test('user is not an admin, displayEditButtons not passed.', () => {
      mockSessionImplementation()
      const { container, getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <SectorCard sector={sectorWithParentSector} />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Sector')
      expect(queryByTestId(EDIT_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(queryByTestId(DELETE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('user is not an admin, displayEditButtons passed.', () => {
      mockSessionImplementation()
      const { container, getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <SectorCard
            sector={sectorWithParentSector}
            displayEditButtons
          />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Sector')
      expect(queryByTestId(EDIT_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(queryByTestId(DELETE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('user is an admin, displayEditButtons not passed.', () => {
      mockSessionImplementation(true)
      const { container, getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <SectorCard sector={sectorWithParentSector} />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Sector')
      expect(queryByTestId(EDIT_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(queryByTestId(DELETE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('user is an admin, displayEditButtons passed.', () => {
      mockSessionImplementation(true)
      const { container, getByTestId } = render(
        <CustomMockedProvider>
          <SectorCard
            sector={sectorWithParentSector}
            displayEditButtons
          />
        </CustomMockedProvider>
      )
      waitForAllEffects()
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Sector')
      expect(getByTestId(EDIT_BUTTON_TEST_ID)).toBeInTheDocument()
      expect(getByTestId(DELETE_BUTTON_TEST_ID)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })
})
