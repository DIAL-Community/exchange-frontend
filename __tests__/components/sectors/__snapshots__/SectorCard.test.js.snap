// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the SectorCard component Should match snapshot - user is an admin, displayEditButtons not passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="border-3 border-transparent"
    data-testid="sector-card"
  >
    <div
      class="flex justify-between border border-dial-gray card-drop-shadow px-4 h-16"
    >
      <div
        class="flex items-center p-4"
      >
        <div
          class="inline-block font-semibold text-button-gray"
        >
          Example Sector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the SectorCard component Should match snapshot - user is an admin, displayEditButtons passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="hover:border-dial-sunshine text-button-gray border-3 border-transparent"
    data-testid="sector-card"
  >
    <div
      class="flex justify-between border border-dial-gray card-drop-shadow px-4 h-16"
    >
      <div
        class="flex-1 cursor-pointer flex items-center p-4"
      >
        <div
          class="inline-flex items-center gap-x-2 inline-block font-semibold text-button-gray"
        >
          Example Sector
          <svg
            data-tooltip-content="Displayable"
            data-tooltip-id="react-tooltip"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
            />
          </svg>
        </div>
      </div>
      <div
        class="inline-flex items-center gap-x-2 p-4"
      >
        <button
          class="flex items-center gap-x-1.5 bg-dial-iris-blue px-2 py-1 rounded-md text-white"
          data-testid="edit-button"
          type="button"
        >
          <img
            alt="Edit"
            height="12px"
            src="/icons/edit.svg"
            width="12px"
          />
          <span
            class="text-sm"
          >
            Edit
          </span>
        </button>
        <button
          class="delete-button gap-x-1.5"
          data-testid="delete-button"
          type="button"
        >
          <svg
            class="text-sm"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z"
            />
            <path
              d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
              fill-rule="evenodd"
            />
          </svg>
          <span
            class="text-sm"
          >
            Delete
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the SectorCard component Should match snapshot - user is not an admin, displayEditButtons not passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="border-3 border-transparent"
    data-testid="sector-card"
  >
    <div
      class="flex justify-between border border-dial-gray card-drop-shadow px-4 h-16"
    >
      <div
        class="flex items-center p-4"
      >
        <div
          class="inline-block font-semibold text-button-gray"
        >
          Example Sector
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the SectorCard component Should match snapshot - user is not an admin, displayEditButtons passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="hover:border-dial-sunshine text-button-gray border-3 border-transparent"
    data-testid="sector-card"
  >
    <div
      class="flex justify-between border border-dial-gray card-drop-shadow px-4 h-16"
    >
      <div
        class="flex-1 cursor-pointer flex items-center p-4"
      >
        <div
          class="inline-flex items-center gap-x-2 inline-block font-semibold text-button-gray"
        >
          Example Sector
        </div>
      </div>
    </div>
  </div>
</div>
`;
